### PR TITLE
Reduce read and write requests and reuse header buffers

### PR DESCRIPTION
**Pull request by derekbit**
- Add -O2 to Makefile
- Reduce read and write requests and reuse header buffers

https://github.com/longhorn/longhorn/issues/4133

### CODE REVIEW

**Comment by innobead** [comment]
cc @shuo-wu @joshimoo @PhanLe1010 @c3y1huang

File: src/longhorn_rpc_client.c [hunk 1]
```diff
@@ -27,13 +27,13 @@ int send_request(struct lh_client_conn *conn, struct Message *req) {
...
         pthread_mutex_lock(&conn->mutex);
-        rc = send_msg(conn->fd, req);
+        rc = send_msg(conn->fd, req, conn->request_header, conn->header_size);
         pthread_mutex_unlock(&conn->mutex);
         return rc;
 }
 
 int receive_response(struct lh_client_conn *conn, struct Message *resp) {
-        return receive_msg(conn->fd, resp);
+        return receive_msg(conn->fd, resp, conn->response_header, conn->header_size);
```
Comment (shuo-wu) [inline src/longhorn_rpc_client.c:36]: Unlike `send_request `, there is no lock protecting `conn->response_header`. Is that OK?

Comment (shuo-wu) [reply src/longhorn_rpc_client.c:36]: Or should/can we malloc the header in `process_request` or `response_process`

Comment (derekbit) [reply src/longhorn_rpc_client.c:36]: > Unlike send_request , there is no lock protecting conn->response_header. Is that OK?

`response_process` is single thread, so it is feasible. @keithalucas is it correct?

> Or should/can we malloc the header along with the req/resp in process_request/response_process?

malloc in `response_process` is ok, but `process_request` is called while receiving each request.

**Review by shuo-wu — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 CC=gcc
-CFLAGS=-c -Wall
+CFLAGS=-O2 -c -Wall
```
Comment (innobead) [inline Makefile:2]: ref: https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

**Comment by derekbit** [comment]
I've run the performance benchmark.
The performance enhancement is mainly in read operations because the reused buffers.
Read IOPS is increased by 5% and bandwidth is increased by 2%.

**Comment by derekbit** [comment]
We can do more stress tests before merging this PR.

**Review by innobead — APPROVED** [review]
LGTM

**Comment by innobead** [comment]
@derekbit Please help use this in longhorn-engine/instance-manager repo for the master, 1.3.x, and 1.2.x. Thanks.

**Pull request merged**